### PR TITLE
Add a HzClientDataLinkFactory to avoid excessive Hz client creation [HZ-1433]

### PR DESCRIPTION
**Pull request by orcunc**
When map journal  of a remote cluster is used within a Jet job, each job instance on a member is creating a new hz client instance. The purpose of this PR is having a single data store that can be re-used for all job instances.

Jira : https://hazelcast.atlassian.net/browse/HZ-1433

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

**Comment by orcunc** [comment]
run-lab-run

**Review by TomaszGaweda — CHANGES_REQUESTED** [review]
I would like to see a test that reproduces problem on an older version, but we check it is ok now.

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ClusterMetaSupplierParams.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import com.hazelcast.internal.journal.EventJournalReader;
+import com.hazelcast.jet.core.EventTimePolicy;
+import com.hazelcast.jet.pipeline.ExternalDataStoreRef;
+import com.hazelcast.jet.pipeline.JournalInitialPosition;
+
+import javax.annotation.Nonnull;
+import java.io.Serializable;
+import java.security.Permission;
+
+// T is input type
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ClusterMetaSupplierParams.java:32]: Ok

File: hazelcast/src/main/java/com/hazelcast/datastore/impl/ExternalDataStoreServiceImpl.java [hunk 2]
```diff
@@ -73,7 +73,8 @@ public boolean testConnection(ExternalDataStoreConfig config) throws Exception {
 
     @Override
     public <DS> ExternalDataStoreFactory<DS> getExternalDataStoreFactory(String name) {
-        ExternalDataStoreConfig externalDataStoreConfig = node.getConfig().getExternalDataStoreConfigs().get(name);
+        Map<String, ExternalDataStoreConfig> externalDataStoreConfigs = node.getConfig().getExternalDataStoreConfigs();
+        ExternalDataStoreConfig externalDataStoreConfig = externalDataStoreConfigs.get(name);
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/datastore/impl/ExternalDataStoreServiceImpl.java:77]: It has been changed in another PR by @frant-hartm

File: hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataStoreFactory.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+ * <p>
+ * Set the client XML as property using {@link HzClientDataStoreFactory#CLIENT_XML} as property key or
+ * Set the client YAML as property using {@link HzClientDataStoreFactory#CLIENT_YML} as property key
+ */
+public class HzClientDataStoreFactory implements DataLinkFactory<HazelcastInstance> {
+
+    /**
+     * The constant to be used as property key for XML
+     */
+    public static final String CLIENT_XML = "client_xml";
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataStoreFactory.java:40]: `CLIENT_XML_PROPERTY` and `CLIENT_YML_PROPERTY`?

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ClusterMetaSupplierParams.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+/**
+ *
+ * @param <T> is input type
+ * @param <R> is return type
+ */
+class ClusterMetaSupplierParams<T, R> implements Serializable  {
+
+    private static final long serialVersionUID = 1L;
+
+    // Optional XML to create a HazelcastInstance which is client that is used to connect to cluster members
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ClusterMetaSupplierParams.java:41]: Please convert in-line comments into JavaDosc

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java [hunk 5]
```diff
@@ -344,39 +352,40 @@ private CompletableFuture<? extends ReadResultSet<? extends T>> readFromJournal(
                                  .toCompletableFuture();
     }
 
-    private static class ClusterMetaSupplier<E, T> implements ProcessorMetaSupplier {
+    public static HzClientDataStoreFactory getDataStoreFactory(HazelcastInstance hazelcastInstance, String name) {
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java:355]: DataStore -> DataLink, same in message below

File: hazelcast/src/test/resources/hazelcast-client-test-external.yaml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      - 127.0.0.1:5701
+    socket-options:
+      tcp-no-delay: true
+      keep-alive: true
+      reuse-address: true
+      linger-seconds: 3
+      buffer-size: 128
+    socket-interceptor:
+      enabled: false
+    aws:
```
Comment (TomaszGaweda) [inline hazelcast/src/test/resources/hazelcast-client-test-external.yaml:17]: IIRC all of those are disabled by default, so maybe we can simplify this config to be as minimal as possible?

File: hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataStoreFactory.java (hunk 3 above)
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataStoreFactory.java:40]: The code will be like this

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java (hunk 5 above)
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java:355]: Ok

File: hazelcast/src/test/resources/hazelcast-client-test-external.yaml (hunk 6 above)
Comment (orcunc) [reply hazelcast/src/test/resources/hazelcast-client-test-external.yaml:17]: Ok

**Review by TomaszGaweda — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java [hunk 7]
```diff
@@ -344,39 +352,40 @@ private CompletableFuture<? extends ReadResultSet<? extends T>> readFromJournal(
...
+        if (!(dataLinkFactory instanceof HzClientDataLinkFactory)) {
+            String className = HzClientDataLinkFactory.class.getSimpleName();
+            throw new HazelcastException("Data store factory '" + name + "' must be an instance of " + className);
+        }
+        return (HzClientDataLinkFactory) dataLinkFactory;
+    }
+
+    // Creates ClusterProcessorSupplier per member. Each ClusterProcessorSupplier is given a
+    // list of partitions IDs
+    private static class ClusterMetaSupplier<E> implements ProcessorMetaSupplier {
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java:369]: This lost the type parameter `T`, Constructor takes `ClusterMetaSupplierParams` but without types.

Also, there were some renames of the type parameters, but now they don't match with the public API, it makes it confusing. It's best to avoid such unnecessary changes.

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java [hunk 8]
```diff
@@ -386,24 +395,41 @@ public int preferredLocalParallelism() {
...
         @Override
         public void init(@Nonnull Context context) {
-            if (clientXml != null) {
-                initRemote();
+            // The order is important.
+            // If dataLinkConfig is specified prefer it to clientXml
+            if (dataLinkRef != null) {
+                HzClientDataLinkFactory hzClientDataLinkFactory = getDataStoreFactory(context.hazelcastInstance(),
+                        dataLinkRef.getName());
+                HazelcastInstance client = hzClientDataLinkFactory.getDataLink();
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java:403]: If this is a non-shared client, how is it closed?

**Review by frant-hartm — CHANGES_REQUESTED** [review]

File: hazelcast/src/main/java/com/hazelcast/jet/pipeline/Sources.java [hunk 9]
```diff
@@ -659,6 +659,57 @@ public static <T, K, V> StreamSource<T> remoteMapJournal(
...
+     *               mapName,
+     *               ExternalDataStoreRef.externalDataStoreRef("my-hzclient-store"),
+     *               JournalInitialPosition.START_FROM_OLDEST,
+     *               EventJournalMapEvent::getNewValue,
+     *               predicate
+     *            ));
+     *       }</pre>
+     *      </p>
+     * @since 5.3
+     */
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/jet/pipeline/Sources.java:696]: Ok

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java [hunk 10]
```diff
@@ -344,39 +353,40 @@ private CompletableFuture<? extends ReadResultSet<? extends T>> readFromJournal(
...
-            this.projection = projection;
-            this.initialPos = initialPos;
-            this.eventTimePolicy = eventTimePolicy;
-            this.permissionFn = permissionFn;
+        ClusterMetaSupplier(ClusterMetaSupplierParams clusterMetaSupplierParams) {
+            this.clusterMetaSupplierParams = clusterMetaSupplierParams;
+
+            clientXml = clusterMetaSupplierParams.getClientXml();
+            externalDataStoreRef = clusterMetaSupplierParams.getExternalDataStoreRef();
+
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java:388]: Ok

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java [hunk 11]
```diff
@@ -459,35 +497,49 @@ public Permission getRequiredPermission() {
...
-            this.initialPos = initialPos;
-            this.eventTimePolicy = eventTimePolicy;
+
+            clientXml = clusterMetaSupplierParams.getClientXml();
+            eventJournalReaderSupplier = clusterMetaSupplierParams.getEventJournalReaderSupplier();
+            predicate = clusterMetaSupplierParams.getPredicate();
+            projection = clusterMetaSupplierParams.getProjection();
+            initialPos = clusterMetaSupplierParams.getInitialPos();
+            eventTimePolicy = clusterMetaSupplierParams.getEventTimePolicy();
+            externalDataStoreRef = clusterMetaSupplierParams.getExternalDataStoreRef();
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java]: Ok

File: hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataStoreFactory.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+import static com.hazelcast.jet.impl.util.ImdgUtil.asClientConfig;
+import static com.hazelcast.jet.impl.util.ImdgUtil.asClientConfigFromYaml;
+
+/**
+ * Creates a HazelcastInstance data store.
+ * <p>
+ * Set the client XML as property using {@link HzClientDataStoreFactory#CLIENT_XML} as property key or
+ * Set the client YAML as property using {@link HzClientDataStoreFactory#CLIENT_YML} as property key
+ */
+public class HzClientDataStoreFactory implements DataLinkFactory<HazelcastInstance> {
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataStoreFactory.java]: It was already renamed before.

File: hazelcast/src/test/resources/hazelcast-client-test-external.xml [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+<hazelcast-client xmlns="http://www.hazelcast.com/schema/client-config"
+                  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                  xsi:schemaLocation="http://www.hazelcast.com/schema/client-config http://www.hazelcast.com/schema/client-config/hazelcast-client-config-5.3.xsd">
+    <cluster-name>dev</cluster-name>
+    <network>
+        <cluster-members>
+            <address>127.0.0.1:5701</address>
+        </cluster-members>
+    </network>
+</hazelcast-client>
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/test/resources/hazelcast-client-test-external.xml:11]: Added new line to end

File: hazelcast/src/test/java/com/hazelcast/jet/pipeline/Sources_withEventJournalTest.java [hunk 14]
```diff
@@ -296,6 +364,24 @@ public void remoteMapJournal_withPredicateAndProjectionFn() {
...
+        PredicateEx<EventJournalMapEvent<String, Integer>> p = e -> e.getNewValue() % 2 == 0;
+
+        // When
+        DataLinkRef dataLinkRef = DataLinkRef.dataLinkRef(HZ_CLIENT_EXTERNAL_REF);
+        StreamSource<Integer> source = Sources.remoteMapJournal(
+                mapName, dataLinkRef, START_FROM_OLDEST, EventJournalMapEvent::getNewValue, p);
+
+        // Then
+        testMapJournal_withPredicateAndProjection(map, source);
+
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/test/java/com/hazelcast/jet/pipeline/Sources_withEventJournalTest.java:382]: Ok

File: hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataLinkFactory.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+    /**
+     * The cached client instance
+     */
+    private HazelcastInstance hazelcastInstance;
+
+    @Override
+    public synchronized HazelcastInstance getDataLink() {
+        if (hazelcastInstance == null) {
+
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataLinkFactory.java:60]: Ok

File: hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataLinkFactory.java [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+     */
+    private DataLinkConfig dataLinkConfig;
+
+    /**
+     * The cached client instance
+     */
+    private HazelcastInstance hazelcastInstance;
+
+    @Override
+    public synchronized HazelcastInstance getDataLink() {
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataLinkFactory.java:58]: Ok

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java (hunk 8 above)
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java:403]: The client created by the dataLinkRef is the shared client. 
It accesses the remote cluster to find the number of partitions. It is not closed

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java [hunk 17]
```diff
@@ -413,27 +441,33 @@ private void initLocal(Set<Partition> partitions) {
...
+
+            // Return a new factory per member owning the given partitions
+            return address -> new ClusterProcessorSupplier<>(addrToPartitions.get(address), clusterMetaSupplierParams);
         }
 
         @Override
         public Permission getRequiredPermission() {
+            if (externalDataStoreRef != null) {
+                return null;
+            }
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java]: Ok
https://hazelcast.atlassian.net/browse/HZ-2073

File: hazelcast/src/main/java/com/hazelcast/datastore/impl/ExternalDataStoreServiceImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/datastore/impl/ExternalDataStoreServiceImpl.java]: It was reverted by  Rename External Data Store to Data Link [HZ-2013]

File: hazelcast/src/main/java/com/hazelcast/datastore/HzClientDataStoreFactory.java [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+    public static final String CLIENT_YML = "client_yml";
+
+    // Reference to configuration
+    private ExternalDataStoreConfig externalDataStoreConfig;
+
+    // The cached client instance
+    private HazelcastInstance hazelcastInstance;
+
+    @Override
+    public synchronized HazelcastInstance getDataStore() {
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/datastore/HzClientDataStoreFactory.java]: Changed to double-checked locking as per the other comment in this review

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java (hunk 7 above)
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java]: Added the parameter T. Also renamed all type parameters according to public API.

File: hazelcast/src/main/java/com/hazelcast/datastore/HzClientDataStoreFactory.java [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+        if (hazelcastInstance == null) {
+
+            ClientConfig clientConfig = buildClientConfig();
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/datastore/HzClientDataStoreFactory.java]: Ok

File: hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataLinkFactory.java [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+    public boolean testConnection() {
+        return true;
+    }
+
+    @Override
+    public synchronized void close() throws Exception {
+        if (hazelcastInstance != null) {
+            hazelcastInstance.shutdown();
+            // Assign to null, so that we don't shut it down again
+            hazelcastInstance = null;
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataLinkFactory.java:88]: Can there be calls to `getDataLink`, after we call `close`?

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java [hunk 21]
```diff
@@ -386,24 +396,41 @@ public int preferredLocalParallelism() {
...
                 client.shutdown();
             }
         }
 
+        // Get remotePartitionCount from given HazelcastInstance
+        private void findRemotePartitionCount(HazelcastInstance client) {
+            HazelcastClientProxy clientProxy = (HazelcastClientProxy) client;
+            ClientPartitionService clientPartitionService = clientProxy.client.getClientPartitionService();
+            // The implementation of getPartitionCount is using an AtomicInteger internally. So it is thread-safe
+            remotePartitionCount = clientPartitionService.getPartitionCount();
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java:431]: Just a note: This call might throw `HazelcastClientOfflineException` if the client is configured with the async start. Can it cause any problems?

File: hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataLinkFactory.java (hunk 20 above)
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/datalink/HzClientDataLinkFactory.java:88]: close() is called when the node is shutting down. So I don't expect to see any other call to a DatalinkFactory

File: hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java (hunk 21 above)
Comment (orcunc) [reply hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java:431]: I don't know. The exception was not getting caught since the beginning. 

As a general rule I can say that in case of any exception the processor does not start. So in the scenario that you have described it seems okay to me that the processor fails.

**Review by mdumandag — APPROVED** [review]
OK from the client's PoV
